File: server/src/models/device.py
from datetime import datetime, timezone
from enum import Enum
from typing import Optional

from beanie import Document, Indexed, PydanticObjectId
from pydantic import Field
from pymongo import IndexModel, ASCENDING


class DeviceStatus(str, Enum):
    ONLINE = "online"
    OFFLINE = "offline"
    BUSY = "busy"
    ERROR = "error"
    MAINTENANCE = "maintenance"


class Device(Document):
    device_uid: Indexed(str, unique=True)
    name: Optional[str] = None
    model: str
    ios_version: str
    status: DeviceStatus = DeviceStatus.OFFLINE
    battery_level: Optional[int] = None
    network_type: Optional[str] = None
    wda_url: str
    companion_id: Indexed(str)
    current_apple_id: Optional[PydanticObjectId] = None
    group_ids: list[PydanticObjectId] = Field(default_factory=list)
    current_task_id: Optional[PydanticObjectId] = None
    last_heartbeat: Optional[datetime] = None
    registered_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    metadata: Optional[dict] = None

    # Hardware identifiers
    udid: Optional[str] = None
    serial_number: Optional[str] = None
    imei: Optional[str] = None
    meid: Optional[str] = None
    wifi_mac: Optional[str] = None
    bluetooth_mac: Optional[str] = None

    # Hardware specs
    cpu_architecture: Optional[str] = None
    hardware_platform: Optional[str] = None
    chip_id: Optional[int] = None
    product_type: Optional[str] = None

    # Jailbreak info
    jailbroken: Optional[bool] = None
    jailbreak_type: Optional[str] = None

    class Settings:
        name = "devices"
        indexes = [
            IndexModel([("status", ASCENDING), ("last_heartbeat", ASCENDING)]),
        ]

File: admin/src/stores/device.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import client from '../api/client'
import { wsClient } from '../ws/client'

export interface DeviceItem {
  id: string
  device_uid: string
  name: string | null
  model: string
  ios_version: string
  status: string
  battery_level: number | null
  network_type: string | null
  wda_url: string
  companion_id: string
  current_apple_id: string | null
  current_task_id: string | null
  last_heartbeat: string | null
  registered_at: string | null
  udid: string | null
  serial_number: string | null
  imei: string | null
  meid: string | null
  wifi_mac: string | null
  bluetooth_mac: string | null
  cpu_architecture: string | null
  hardware_platform: string | null
  chip_id: number | null
  product_type: string | null
  jailbroken: boolean | null
  jailbreak_type: string | null
}

export interface DeviceStats {
  total: number
  online: number
  offline: number
  busy: number
  error: number
  online_rate: number
}

export const useDeviceStore = defineStore('device', () => {
  const devices = ref<DeviceItem[]>([])
  const total = ref(0)
  const stats = ref<DeviceStats>({ total: 0, online: 0, offline: 0, busy: 0, error: 0, online_rate: 0 })
  const loading = ref(false)

  async function fetchDevices(params: {
    status?: string
    group_id?: string
    page?: number
    size?: number
    sort?: string
  } = {}) {
    loading.value = true
    try {
      const { data } = await client.get('/devices', { params })
      if (data.code === 0) {
        devices.value = data.data.items
        total.value = data.data.total
      }
    } finally {
      loading.value = false
    }
  }

  async function fetchStats() {
    const { data } = await client.get('/devices/stats')
    if (data.code === 0) {
      stats.value = data.data
    }
  }

  function setupWsHandlers() {
    wsClient.on('device.status_changed', (payload: any) => {
      const idx = devices.value.findIndex(d => d.device_uid === payload.device_uid)
      if (idx >= 0) {
        devices.value[idx].status = payload.new_status
        if (payload.battery_level !== undefined) {
          devices.value[idx].battery_level = payload.battery_level
        }
        if (payload.current_task_id !== undefined) {
          devices.value[idx].current_task_id = payload.current_task_id
        }
      }
    })

    wsClient.on('device.heartbeat', (payload: any) => {
      for (const dev of payload.devices || []) {
        const idx = devices.value.findIndex(d => d.device_uid === dev.device_uid)
        if (idx >= 0) {
          devices.value[idx].status = dev.status
          if (dev.battery_level !== undefined) {
            devices.value[idx].battery_level = dev.battery_level
          }
        }
      }
    })
  }

  return { devices, total, stats, loading, fetchDevices, fetchStats, setupWsHandlers }
})
